hibernate-core/src/main/java/org/hibernate/persister/entity/JoinedSubclassEntityPersister.java

```diff
@@ -1376,6 +1386,29 @@ public void pruneForSubclasses(TableGroup tableGroup, Map<String, EntityNameUse>
 		}
 
 		final List<TableReferenceJoin> tableReferenceJoins = tableGroup.getTableReferenceJoins();
+		if ( needsTreatDiscriminator ) {
+			if ( tableReferenceJoins.isEmpty() ) {
+				// We need to apply the discriminator predicate to the primary table reference itself
+				final String discriminatorPredicate = getPrunedDiscriminatorPredicate( entityNameUses, metamodel, "t" );
+				if ( discriminatorPredicate != null ) {
+					final NamedTableReference tableReference = (NamedTableReference) tableGroup.getPrimaryTableReference();
+					tableReference.setPrunedTableExpression( "(select * from " + getRootTableName() + " t where " + discriminatorPredicate + ")" );
+				}
+			}
+			else {
+				// We have to apply the discriminator condition to the root table reference join
+				boolean applied = applyDiscriminatorPredicate(
+						tableReferenceJoins.get( 0 ),
+						(NamedTableReference) tableGroup.getPrimaryTableReference(),
+						entityNameUses,
+						metamodel
+				);
+				for ( int i = 0; !applied && i < tableReferenceJoins.size(); i++ ) {
+					final TableReferenceJoin join = tableReferenceJoins.get( i );
+					applied = applyDiscriminatorPredicate( join, join.getJoinedTableReference(), entityNameUses, metamodel );
+				}
```

Comment:
I think I'd feel more comfortable if you add an assert at the end, and if it fails, we have to think about how to handle this scenario.

```suggestion
				}
				assert applied;
```